### ISSUE TITLE
Convert method to extension method

### ISSUE THREAD

**Issue by giggio** [code-cracker/code-cracker#365]
This:

````csharp
public class Test 
{
    public char GetThirdCharacter(string str)
    {
        return str[2];
    }
}
````

becomes

````csharp
public static class Test 
{
    public static char GetThirdCharacter(this string str)
    {
        return str[2];
    }
}
````

A class that has extension methods has to be non generic and static. So, the diagnostic should never be raised to methods inside generic classes.

I have concerns I will express on comments bellow, let's see if we can find a nice scenario where we can make this helpful.

**Comment by giggio** [comment]
Bringing in @Kavignon who had the idea on issue #364.

**Comment by giggio** [comment]
Non static classes can be turned static, but only if no method inside the class is accessing `this`.
Also, the only allowed constructor would be the default one, and only if the had no code.
All references to the methods would need to change.

**Comment by giggio** [comment]
I am not sure the best approach is to turn the method and class static AND make it an extension method. Maybe we could work only on static classes? @Kavignon @code-cracker/owners what do you think?

**Comment by baks** [comment]
This is nice :+1: 
For me better fits something like "Extract as extension method".

Usually when I work on something, the method that can be turned to extension method, first show up as `private static` method in some other class . And after a while, idea came up ->  this can be extension method.

I have feeling that this is common way how extension method are created.

**Comment by giggio** [comment]
@baks This is a nice idea, we could work on a new issue for "Extract as extension method". What do you think?

**Comment by Kavignon** [comment]
I think its a wonderful idea. Could not have said that in a better way.

**Comment by giggio** [comment]
I just realized that to be able to "extract as extension method" we need a selection. We are currently doing refactorings as hidden diagnostics + code fix, so "extract as extension method" would only work if we did a real refactoring. The problem is that we can't deploy refactorings with nuget yet (this is a roslyn thing).

**Comment by Kavignon** [comment]
Well, I dont think that the methods need to be private before they can be extracted as an extension. We could just look if all methods are static, that they are extending the same type in the the class which needs to be static.

**Comment by Kavignon** [comment]
Actually I've describe an extension class. Nonetheless, I do think we don't have to look for the method accessor

**Comment by baks** [comment]
@giggio correct me if I'm wrong. 
If we detect a method that can be considered as extension method and mark that as hidden diagnostics and propose code fix - `Extract method {methodName} to extension method`. It won't work in that way?

**Comment by giggio** [comment]
@baks No, because we wouldn't know what to extract. One line? Two lines?
That is the difference of a refactoring, one of the parameters is the user selection.

**Comment by baks** [comment]
@giggio I see your point. 
But if we just propose to extract entire method body to extension method?
Maybe I didn't explain it well but my thoughts were to convert (extract) method to extension method. If a method can be considered as extension method, create new static class, move that method to this class and make it extension method.

```csharp
class SomeService
{
    public string GetValue()
    { 
          return GetThirdCharacter("");
    }

    private static char GetThirdCharacter(string str)
    {
        return str[2];
    }
}
```

becomes

```csharp
class SomeService
{
    public string GetValue()
    { 
         return "".GetThirdCharacter();
    }
}

public static class StringExtensions
{
    public static char GetThirdCharacter(this string str)
    {
        return str[2];
    }
}
```

**Comment by giggio** [comment]
@baks I created #373 for your suggestion, which is different from this one. I like it!
This issue is similar to #373, but different. In #373 we extract the method to a different class, and on this one we change the current class to become a static class.
